File: PPS/06/7.c
/* binary search */

// 1. Take num as the number to be searched
// 2. set c to 0
// 3. set lp = 0 and up = length of array
// 4. if lp is more than up, go to step 7
// 4. mid = (lp+up)/2
// 5. if arr[mid] is equal to num,  increase the value of c and go to step 7
// 6. if arr[mid] is greater than num, up = mid-1 and go to step 4
// 6. if arr[mid] is less than num, lp = mid-1 and go to step 4
// 7. if c is 1, the number is present
// 8. if c is 0, the number is not present

#include<stdio.h>

int main()
{
    int arr[20],n,num,c=0;
    printf("Enter the size of array : ");
    scanf("%d",&n);
    printf("Enter the values : ");
    for(int i=0;i<n;i++)
        scanf("%d",&arr[i]);
    printf("Enter the number to be searched : ");
    scanf("%d",&num);
    int lp=0,up=n;
    while(lp<=up)
    {
        int mid = (lp+up)/2;
        if(arr[mid]==num)
        {
            c++;
            break;
        }
        else if(arr[mid]>num)
            up=mid-1;
        else
            lp=mid+1;
    }
    if(c==1)
        printf("The number is present.\n");
    else
        printf("The number is not present.\n");
}

File: PYTHON/7/7.py
# linear or binary

def linear(tup,num):
    for i in tup:
        if i == num:
            return 1

def binary(tup,num):
    lp =0
    up = len(tup)-1
    while lp<=up:
        mid = (lp+up)//2
        if tup[mid]==num:
            return 1
        elif tup[mid]>num:
            up = mid-1
        else:
            lp=mid+1
    
tup = (1,2,3,4,5,6,7,8,9)
num = int(input('Enter the number to be searched : '))
c = input("Enter 'l' for linear and 'b' for binary : ")
if c == 'l':
    count = linear(tup,num)
if c=='b':
    count = binary(tup,num)
if count == 1:
    print('Search Successful')
else:
    print('Search Unsucessful')

